src/python/WMCore/MicroService/MSPileup/MSPileupTasks.py

```diff
@@ -0,0 +1,450 @@
+    logger = spec.get('logger', getMSLogger(False))
+    rucioClient = spec['rucioClient']
+    rucioAccount = spec['rucioAccount']
+    report = spec['report']
+    marginSpace = spec.get('marginSpace', 0)
+
+    # extract relevant part of our pileup document we'll use in our logic
+    expectedRSEs = set(doc['expectedRSEs'])
+    currentRSEs = set(doc['currentRSEs'])
+    pileupSize = doc['pileupSize']
+    pname = doc['pileupName']
+    inputArgs = {'scope': 'cms', 'name': pname, 'account': rucioAccount}
+    modify = False
+    localCopyOfCurrentRSEs = list(currentRSEs)
+
+    # if expectedRSEs is different than currentRSEs, then further data placement is required
+    # it's possible that data removal is required!
+    if expectedRSEs != currentRSEs:
+        msg = f"active task {uuid}, further data placement required for pileup name: {pname}, with expectedRSEs: {expectedRSEs} and data currently available at: {currentRSEs}"
+        logger.warning(msg)
+
+        # get list of replication rules for our scope, pileup name and account
+        if isinstance(rucioClient, Client):
+            rules = [r for r in rucioClient.list_replication_rules(inputArgs)]
+        else:
+            rules = [r for r in rucioClient.listDataRules(pname)]
+        # for each rse_expression in rules get list of rses
+        for rdoc in rules:
+            if isinstance(rucioClient, Client):
+                rses = [r['rse'] for r in rucioClient.list_rses(rdoc['rse_expression'])]
+            else:
+                rses = [rse for rse in rucioClient.evaluateRSEExpression(rdoc['rse_expression'])]
+            rdoc['rses'] = rses
+
+        # for each rule matching the DID + Rucio account
+        for rdoc in rules:
+            rid = rdoc['id']
+            rses = rdoc['rses']
+            for rse in rses:
+                # if rule RSE is not in expectedRSEs, then this rule needs to be deleted
+                if rse not in expectedRSEs:
+                    # upon successful rule deletion, also remove the RSE name from currentRSEs
+                    if isinstance(rucioClient, Client):
+                        rucioClient.delete_replication_rule(rid)
+                    else:
+                        rucioClient.deleteRule(rid)
+                    if rse in doc['currentRSEs']:
+                        doc['currentRSEs'].remove(rse)
+                    msg = f"rse {rse} rule is deleted and remove from currentRSEs"
+                    logger.info(msg)
+                    # delete rid in ruleIds (if any)
+                    if rid in doc['ruleIds']:
+                        doc['ruleIds'].remove(rid)
+                        msg = f"rule Id {rid} is removed from ruleIds"
+                        logger.info(msg)
+                else:
+                    # else, save this rule RSE in the local copy of currentRSEs.
+                    # This rule is likely still being processed by Rucio.
+                    if rse not in doc['currentRSEs']:
```

Comment:
Given that this rule might be either in progress or satisfied, we should append it to the local copy instead (localCopyOfCurrentRSEs). Which will then be used in the coming lines to decide which rules are still really missing (different than: which rules are either missing or in progress).

Reply:
corrected.